baybe/constraints/base.py

```diff
@@ -116,6 +119,62 @@ class ContinuousConstraint(Constraint, ABC):
     eval_during_modeling: ClassVar[bool] = True
     # See base class.
 
+    _numerical_only: ClassVar[bool] = True
+    # See base class.
+
+
+@define
+class CardinalityConstraint(Constraint, ABC):
+    """Abstract base class for cardinality constraints.
+
+    Places a constraint on the set of nonzero (i.e. "active") values among the
+    specified parameters, bounding it between the two given integers,
+        ``min_cardinality`` <= |{p_i : p_i != 0}| <= ``max_cardinality``
+    where ``{p_i}`` are the parameters specified for the constraint.
+
+    Note that this can be equivalently regarded as L0-constraint on the vector
+    containing the specified parameters.
+    """
+
+    min_cardinality: int = field(default=0, validator=[instance_of(int), ge(0)])
+    "The minimum required cardinality."
+
+    max_cardinality: int = field(validator=instance_of(int))
+    "The maximum allowed cardinality."
+
+    _numerical_only: ClassVar[bool] = True
+    # See base class.
+
+    @max_cardinality.default
+    def _default_max_cardinality(self):
+        """Use the number of involved parameters as the upper limit by default."""
+        return len(self.parameters)
+
+    def __attrs_post_init__(self):
+        """Validate the cardinality bounds.
+
+        Raises:
+            ValueError: If the provided cardinality bounds are invalid.
+            ValueError: If the provided cardinality bounds impose no constraint.
+        """
+        if self.min_cardinality > self.max_cardinality:
+            raise ValueError(
+                f"The lower cardinality bound cannot be larger than the upper bound. "
+                f"Provided values: {self.max_cardinality=}, {self.min_cardinality=}."
+            )
+
+        if self.max_cardinality > len(self.parameters):
+            raise ValueError(
+                f"The cardinality bound cannot exceed the number of parameters. "
+                f"Provided values: {self.max_cardinality=}, {len(self.parameters)=}."
+            )
+
+        if self.min_cardinality == 0 and self.max_cardinality == len(self.parameters):
```

Comment:
Re1: The intent here is try to reduce the number of cardinality constraint when possible, because the cardinality constraint will be relaxed to a nonlinear inequality constraint and feed to the botorch optimizer in the upcoming PR. Based on my first assessment, such a nonlinear inequality constraint is computationally heavy and will slow down the optimization. 
Re2: Good catch, will fix it. Thanks.